### DISCUSSION TITLE
Custom Fields API

### DISCUSSION THREAD

**Discussion by dantium** [medusajs/medusa#1947] [Feature Requests]
Comment from @olivermrbl:
### What
In Q1 2023, we will revisit our Product API and begin to extend its capabilities. In the first iteration, we will introduce Nested Categories (see RFC (TBA)), and later we will add support for custom fields. ~~We are likely to start scoping the Custom Fields API in mid-Q1.~~

-----
**OP**
I checked the admin demo and could see no way to add custom attributes for products. Did I miss something or would this need to be done via code?

**Comment by synapseit** [comment]
Any update on this?
I saw that "Extended Product API (custom fields, publishing control, and more)" is part of 2022 plan. In the meanwhile, what are some possible workaround?
Does it make sense to create a CustomProduct entity with relevant Services and Endpoints? Any other advice?

**Reply by olivermrbl** [reply]
We are yet to decide on an approach for supporting custom attributes on entities in Medusa. There are a couple of different approaches to take, and they all come with pros and cons, and we haven't settled on the right tradeoff. 

To briefly touch upon the approaches currently in consideration:
- Adding `custom_attributes` field of type JSONB to relevant entities
- Adding a new generic entity `CustomProperty` that references any resource (e.g. product) and holds a key-value pair
- Allowing you to extend the existing database schema with custom properties

There are likely more options, which will be uncovered during the technical design phase. We'll include the community in the decision as soon as we start working on the feature. 

Until then, you can either use the existing metadata field, add a new entity (option 2 from above) that holds the customer attributes for your product, or use the Medusa Extender.

**Reply by sergioruizsan** [reply]
Hi @olivermrbl, I do have few questions:

A. Will it be possible to filter products by these custom fields? 
B. Will the custom fields be fixed for all the products (like the current Product Details)?
C. Will the field values be unique? So, for example, if you are adding a second product which you want to have a similar value, will it show a dropdown with the already available values or, alternatively, the ability to create a new one?

The reason I am asking is because our team wanted to use the variant options to filter out products for a search page and we potentially think that these custom fields could be helpful. A search page that would have a side menu with:
1. All the different keys (for example: brand, material, etc) and all possible values (for example: brand-zara, brand-superdry,... and material-cotton, material-synthetic,...). 
2. A slider to filter by price
3. A slider to filter by percent of cotton of a clothing

Ideally, for point 1, tags could work as well, but those tags don't fall into a category. And a workaround we thought about is to create a tag like `[category]:[value]`, for example `brand:zara`. However they are not seen as unique values by the Admin UI, I mean there is no dropdown to select an existing one. Also, the list of products based on tags can only be retrieved from the Admin API.
For point 2 and 3 there is nothing in place that allow us to do that as far as I've seen. Correct me if I am wrong please.

Thanks 😄

**Reply by olivermrbl** [reply]
Thanks for chiming in @sergioruizsan. We are yet to start the technical design process, so I will only share our high-level thoughts on the feature to avoid empty promises.

The Custom Fields API is likely to be constrained to the following functionality (at least in the 1st iteration):
- All entities can be extended with custom properties
- Extending an entity permanently adds the property to all entities of its type
- Extending an entity does not alter the core database table, but instead, the [EAV pattern](https://pbedn.github.io/post/2020-05-25-entity-attribute-value/) will be applied (TBD)
- Custom properties are filterable
- Adding custom properties can be done in several ways; in the simple case, you include it in `medusa-config.js`, and we take of it for you under the hood:
```
modules.export = {
  plugins: [ ... ],
  customFields: {
    product: [{
	name: "vendor_id",
	type: "string"
    }]
  }
}
```

As mentioned, I can't share implementation specifics around the EAV pattern nor the full developer experience yet, as we are still in planning.

-------

For your scenario 1., you might be better off with a feature soon to be released. We are working on an Advanced Product Category API, allowing you to create a nested hierarchy of categories. I imagine you could have `brand > zara`, `brand > superdry`, `brand > h&m`, etc. The same goes for material. Products would be filterable by these categories.

**Reply by mariusa** [reply]
> Adding custom properties can be done in several ways; in the simple case, you include it in medusa-config.js

This won't work for multi-tenant. Better to have one solution which doesn't put more roadblocks on  multi-tenant.
For example https://shopify.dev/apps/custom-data/metafields/definitions/manage-metafield-definitions

Please also consider https://help.shopify.com/en/manual/custom-data/metaobjects

**Reply by sergioruizsan** [reply]
Thanks for your quick reply @olivermrbl 😄 . Sounds good. We will give it a try with the Category API then. Looking forward for the Custom Fields API! 🚀

**Comment by madsviktor** [comment]
@olivermrbl Hey Oliver, is there any update on a timeline?

Related, is it correct that there's no way to edit metadata in medusa admin?

Thanks,

**Reply by olivermrbl** [reply]
Thanks for pitching in @madsviktor. This is on our roadmap, and we plan to start the technical design process within the next couple of months. I can't give you an exact date of release, but a _guesstimate_ would be end of Q1 or start Q2 :)

**Reply by madsviktor** [reply]
Thanks for getting back so quick :) Sounds great, thanks!

**Comment by DaoQuangThiet** [comment]
Hi medusajs team, now i want to add custom field "description" for collection and many more custom fields. so how to do it?
![image_2023_03_10T03_26_49_295Z](https://user-images.githubusercontent.com/67154017/224215772-7ba2e173-7518-4a56-b801-5722962e0ad1.png)

**Reply by flashjames** [reply]
> Hi medusajs team, now i want to add custom field "description" for collection and many more custom fields. so how to do it? ![image_2023_03_10T03_26_49_295Z](https://user-images.githubusercontent.com/67154017/224215772-7ba2e173-7518-4a56-b801-5722962e0ad1.png)

This is not possible right now, but should probably be possible when Custom Fields API is out.

I settled for adding the data to metadata field for now.

**Comment by javialgaba** [comment]
@olivermrbl Hi Oliver, I do have a few questions regarding this feature. My team is considering using MedusaJS, but we've found a roadblock, and potentially, Custom Fields API, could be the answer to that problem. We would like to use MedusaJS for a restaurant, but the examples we've found so far assume every product is a dish, which is not our case (https://medusajs.com/blog/next-js-storefront-food-ecommerce-restaurant-medusa-paystack/). For our needs, we will need to have the concept of a modifier (extra cheese, ketchup, etc): it's similar to a variant but can be shared between different products. On the contrary, a variant must be defined upfront and cannot be shared with other products. The metadata field is not optimal, because if for instance, the restaurant runs out of stock, we will need to iterate through all the products just to find out which ones have that modifier and set it as out of stock, for instance. Custom Fields seem like a very promising feature because it could allow extending Product entity without forking Medusa. The other solution would be to use Medusa Extender as far as I know. Could Custom Fields fit our needs? Thank you very much.

**Reply by IoannisFlo** [reply]
hey @javialgaba I'm new to the codebase so probably mistaken here, but wouldn't [Product Categories](https://docs.medusajs.com/api/admin#tag/Product-Categories) help you accomplish your goal here? it was shipped in the latest release (version 1.8), there's a [blog post about it](https://medusajs.com/blog/announcing-product-categories/). 

not sure if you'd still need custom fields for your use case

**Comment by teleputer** [comment]
I have been waiting over a year for custom product fields, and it keeps getting pushed back.  It was hard for me to read that the technical design has not even been started.  Strapi has done a great job at this and Medusa should work / be designed very similar. 
We need custom fields where we can define a field by string, decimal, integer, boolean etc...   

Thank you!

**Reply by olivermrbl** [reply]
I am curious to hear where our current extensibility capabilities fall short. Would you care to elaborate?

We allow you to extend entities, services, repositories, and much more. So even though we do not have a dedicated Custom Fields API, our engine meets the fundamental requirements of what such an API would offer to developers.

**Reply by teleputer** [reply]
Hi @olivermrbl - We use dedicated fields that contain text, integer, binary, boolean etc..  and use these fields for displaying custom product features based on the logic of product values, along with integrations into 3rd party applications like Algolia for searching sizes and weights etc...  However, your answer above give me little doubt that you understand ecommerce and this concerns me more than not having custom fields.  This, along with height, length, width, and weight are integer and not decimal tell me you do not understand ecommerce.  
I was very excited when MedusaJS first started but I have to say I think you are heading in the wrong direction for the masses to use and support this application at a level that it is intended to be used at.  I wish you the best of luck but as of now, I am no longer considering MedusaJS as a solution for my clients.
------**
Custom fields in an ecommerce application can be very useful for a number of reasons:

Gathering additional customer information: Custom fields allow you to collect specific information about your customers that may not be included in standard ecommerce fields, such as dietary restrictions or preferred shipping methods. This information can be useful for personalizing the customer experience and providing targeted marketing messages.

Tailoring product information: Custom fields can also be used to provide additional information about your products that might not be included in the standard product fields. For example, if you sell clothing, you could create custom fields for material type or care instructions.

Supporting complex products: If you sell products that require more information than standard products, custom fields can help you manage these complexities. For example, if you sell custom-made furniture, you might use custom fields to gather specific measurements or design preferences from your customers.

Integrating with other systems: Custom fields can also be used to integrate your ecommerce application with other systems, such as your CRM or accounting software. By capturing specific data in custom fields, you can ensure that this information is automatically synced with other systems in your organization.

Overall, custom fields can be a powerful tool for tailoring your ecommerce application to the specific needs of your business and customers.

**Reply by olivermrbl** [reply]
Hi @teleputer 
 
I appreciate your engagement with our open discussions. Though, I want to remind you to please keep a positive and respectful tone, as this is crucial in building a healthy and collaborative open-source community.

The reason we have public roadmaps and feature discussions is to gather constructive feedback and ideas for building the best possible features. With that in mind, I thank you for your input and would love to learn more to understand how we can improve our product and documentation around some of these concepts.

As I mentioned in my first message, albeit we don't have a Custom Fields API, we support many of the use cases you highlighted here at the fundamental level. This means it might not be as intuitive and straightforward as you'd like it to be (and as Strapi offers it), but it is possible, which is what I want to stress.

There are two ways of going about it; extending entities or using metadata. 

Extending existing database schemas can be achieved with a few steps and a portion of custom code in your Medusa project. Let's take your example of adding additional customer information: 

```ts
import { Column, Entity } from "typeorm"

import {
  // alias the core entity to not cause a naming conflict
  Customer as MedusaCustomer,
} from "@medusajs/medusa"

@Entity()
export class Customer extends MedusaCustomer {
  @Column()
  dietary_restrictions: string[]
}
```

Or custom product information:
```ts
import {
  // alias the core entity to not cause a naming conflict
  Product as MedusaProduct,
} from "@medusajs/medusa"

@Entity()
export class Product extends MedusaProduct {
  @Column()
  care_instructions: string
}
```

In both cases, these fields will be present on your database entities after running the related migrations. Additionally, they will also be exposed to all standard CRUD operations through the API. 

You can read more about [this approach in our documentation](https://docs.medusajs.com/development/entities/extend-entity).

If you want to avoid extending the database schemas, there's an alternative and more well-known customization approach: using metadata. Almost all entities in Medusa contain a metadata field that is modifiable directly via our API and from within Medusa Admin.

We've recently added UI metadata forms to many domains. See this tweet for a quick demo: 
https://twitter.com/medusajs/status/1643304287777218571?s=20


And to illustrate with one of the cases from before, the metadata on customers would look something like this:
```ts
{
  "customer": {
    "id": "cus_01G2SG30J8C85S4A5CHM2S1NS2",
    "email": "user@example.com",
    "first_name": "Elon",
    "last_name": "Musk",
    "billing_address_id": "addr_01G8ZH853YPY9B94857DY91YGW",
    "created_at": "2019-08-24T14:15:22Z",
    "updated_at": "2019-08-24T14:15:22Z",
    "deleted_at": "2019-08-24T14:15:22Z",
    "metadata": {
      "dietary_constraints": "Lactose intolerant"
    }
  }
}
```

And the UI like this:

<img width="1006" alt="CleanShot 2023-04-20 at 10 27 41@2x" src="https://user-images.githubusercontent.com/59018053/233306942-c4bbdd3e-8343-4da5-9fe1-98fb44533110.png">

I hope this helps explain the current extensibility capabilities of Medusa. Let me know if you have any additional questions about these two approaches. I am more than happy to elaborate.

There's actually also a third option, which would be to integrate Medusa with Strapi. Although we don't have an official integration, there's a [community plugin available](https://github.com/SGFGOV/medusa-strapi-repo). I recommend checking that out and potentially joining our Discord, as you'll be able to speak directly with the maintainers there.

And lastly, to answer your comment on the other question, the reason we have lowered the priority on a Custom Fields API is that we already have these mechanisms in place for adding custom data to entities. We would rather spend time tackling some of the bigger opportunities, like bringing commerce logic to serverless and edge environments.

**Reply by teleputer** [reply]
I apologize for being disrespectful.  I have the utmost respect for you and everyone working on MedusaJS.  You have some brilliant and talented people working on this product. 
I was just frustrated when you posted yesterday, "We do not plan to build a dedicated Custom Fields API shortly, as we have other features in the roadmap (to be shared soon) that take higher priority."  I was led to believe from previous posts that this was a priority in the roadmap, especially with the number of upvotes. 
I understand extending the database fields and using metadata, and yes, it most likely will be helpful to some but not entirely to others.  However, this solution adds more time, code, and other complexities and migrations for adding a custom field.  The approach you gave above is ideal for services but more of a workaround for front-end custom-facing updates. 
When you have to work with other applications which already provide custom fields, like Strapi, it is frustrating to switch between the two applications where something is straightforward to do in one but time-consuming and different to do in another.  
I was hoping that Strapi and Medusa would be a viable solution that was easy to integrate and work together, and this is not your fault but my expectations.   My requirements are different from others.  I need a headless CMS and eCommerce system that easily integrates and works together to monetize products and/or content.

**Reply by modzincltd** [reply]
Just incase it helps anyone this can be easily shown on the store front like so 👍 
`{customer.metadata.dietary_constraints}`

**Comment by empz** [comment]
With Medusa 1.8 is now possible to simply extend an entity to add new fields, right? So is this the recommended approach going forward or are you going to also provide a "customFields" configurable in the medusa-config ?

**Reply by olivermrbl** [reply]
Yes, this is possible. We do not plan to build a dedicated Custom Fields API in the near future, as we have other features in the roadmap (to be shared soon), that take higher priority.

**Reply by teleputer** [reply]
Custom fields has over 70+ upvotes on this Github discussion.  No other discussion for MedusaJS has as many up votes than this topic.  How can this not be a high priority on the road map?  How are you creating your map?    Please, copy what Strapi has done for custom fields and use this as the technical design.  They have done a great job at implementing this feature for CMS.

**Reply by ivanushkaPr** [reply]
Great, but extending entites without the ability to extend create product modal leads us to the need to create your own crud panel. And that kinda sucks.

**Comment by matvance** [comment]
After spending over 5 hours testing medusa js, I'm sad to say that the lack of a custom fields API is too much of a worry to use the library, so I won't be using it to build my e-commerce site.
You are doing the great job guys but there is much work left. Or the library is supposed to be used for small e-shops only? Either way I am not building my e-commerce on it as for now.

**Reply by olivermrbl** [reply]
Hi @komarnisky – obviously, I am sorry to hear, that you deemed Medusa insufficient for your use case. 

I would love to understand your needs and where our current extensibility capabilities fall short. That way, we can take that into consideration in our roadmap planning in the future. 

Please see my [other comment](https://github.com/medusajs/medusa/discussions/1947#discussioncomment-5657543) on the topic.

**Reply by matvance** [reply]
Well, adding a custom fields to search by in products listing later on is a must-have to build a bigger e-commerce website with more than a couple of categories. Also, I have noticed that the product description field is just a plain text while in the other engines you can add images etc.
Do you think I might solve these problems with strapi integration?

**Reply by AlexTechNoir** [reply]
> Do you think I might solve these problems with strapi integration?

I tried to install medusa-plugin-strapi some time ago, but it wasn't working properly. Strapi plugin is supported by community, not by Medusa maintainers. From what I saw, it's not being updated regularly and probably will work only with older Medusa versions.

Creating your own entities and APIs seem to be the way to implement _some_ of the desirable features, so far. I took [this comment](https://github.com/medusajs/medusa/discussions/2228#discussioncomment-3707069) as a starting point

**Reply by khqnn** [reply]
> After spending over 5 hours testing medusa js, I'm sad to say that the lack of a custom fields API is too much of a worry to use the library, so I won't be using it to build my e-commerce site. You are doing the great job guys but there is much work left. Or the library is supposed to be used for small e-shops only? Either way I am not building my e-commerce on it as for now.

I was able to add custom field in the Product entity but I am unable to filter products based on this custom field. The custom field is "business_id" and when I try to filter products I got the following error,
```
{
    "type": "invalid_data",
    "message": "property business_id should not exist"
}
```

Whereas the "business_id" is presented in the database schema

**Reply by olivermrbl** [reply]
You will have to extend the API validators to accept your added properties. See more in [this comment](https://github.com/medusajs/medusa/discussions/3965#discussioncomment-5805367).

**Comment by Phawhan** [comment]
Hi Team 
I have added custom attribute called "store_name" to the product entity and created loader attached below to include the store_name in product response, now I am getting the store_name attribute in the Store/products api response but not in admin/products api  and also wanted to assign a value to the custom attributes while creating the products.
export default async function(){

    const imports = await import (
        "@medusajs/medusa/dist/api/routes/store/products/index"
    ) as any
    imports.allowedStoreProductsFields = [
        ...imports.allowedStoreProductsFields,
        "store_name",
      ]
      imports.defaultStoreProductsFields = [
        ...imports.defaultStoreProductsFields,
        "store_name",
      ]
}

Thanks in advance

**Comment by mehmetuken** [comment]
Hey @olivermrbl,

Adding the custom attributes feature, similar to the one in Saleor, would be much more logical. By assigning these attributes to the product type, as soon as the product type is selected, those attributes would be available, creating a rapid product attributes process. Attributes can have multiple options such as free text, multiple-choice, multiple-selection, checkboxes, and more.

**Comment by Phawhan** [comment]
Hey @olivermrbl 

I have few doubts regarding medusajs, if you help it will be better.

How can we find which micro-services(functions) are bind to which api, lets suppose considering an Admin Product POST api that creates a product with variants will use ProductService functions , VariantService functions and many more. How to find all the services that are bind with the API.

And How can we extend the product variants with custom attributes in create-product api

Product create api validator
export declare class AdminPostProductsReq {
    title: string;
    subtitle?: string;
    description?: string;
    is_giftcard: boolean;
    discountable: boolean;
    images?: string[];
    thumbnail?: string;
    handle?: string;
    status?: ProductStatus;
    type?: ProductTypeReq;
    collection_id?: string;
    tags?: ProductTagReq[];
    sales_channels?: ProductSalesChannelReq[];
    categories?: ProductProductCategoryReq[];
    options?: ProductOptionReq[];
    variants?: ProductVariantReq[];
    weight?: number;
    length?: number;
    height?: number;
    width?: number;
    hs_code?: string;
    origin_country?: string;
    mid_code?: string;
    material?: string;
    metadata?: Record<string, unknown>;
}

variants validator
declare class ProductVariantReq {
    title: string;
    sku?: string;
    ean?: string;
    upc?: string;
    barcode?: string;
    hs_code?: string;
    inventory_quantity?: number;
    allow_backorder?: boolean;
    manage_inventory?: boolean;
    weight?: number;
    length?: number;
    height?: number;
    width?: number;
    origin_country?: string;
    mid_code?: string;
    material?: string;
    metadata?: Record<string, unknown>;
    prices: ProductVariantPricesCreateReq[];
    options?: ProductVariantOptionReq[];
}

Let me know for any more information regarding above, Thanks in advance

**Comment by Rohit041299** [comment]
Hi @olivermrbl 
I am a fresher in medusa js, read the documentation still a little bit confuse.
Can you explain me the hierarchy structure of how interface, models, services etc are interacting in Medusa js.

Thanks.

**Comment by drennvinn** [comment]
Hi!! I am also a new medusa user and It would be really great if custom fields could be added to Medusa because it's true that it's a task we would like to get rid of quickly in our projects. Just now, I would like to add an image field to product-categories, and it's a lot of code for something that could be directly managed from the Medusa admin or via a JSON file...
I've been working with Directus for a while, and they have an excellent custom entity management system. In addition to custom entities and fields, there are 'system collections' that remain unchanged to keep the core stable, but you can still add additional custom fields to them.
I hope I can contribute to the project once I become more familiar with medusa 😁

**Reply by TrinhQuangvo** [reply]
have you found some way to add an image to prod_cate yet?

**Comment by vholik** [comment]
I created the plugin that adds custom attributes to the products. If you're looking for a good solution definetely [check it out](https://www.npmjs.com/package/medusa-custom-attributes). Also there is a [blog post](https://rigby.pl/en/blog/custom-attributes-in-medusa.js) how to use that plugin.

**Reply by AkshatKhamesra** [reply]
We tried this and the custom attribute is getting added but we are unable to find it in the database, can you help guide how and where to find it?

**Reply by vholik** [reply]
Yeah, I can help you with that. If you added single, multiple and boolean attributes you can find it by "attribute_value" column. If you added range attribute you can find it by "int_attribute_value" column in a db.

**Reply by Vidisha-Khalpada** [reply]
Hello, I had one question. How can I migrate the old data using the seed command of medusa if I use your custom attribute? Example if I use seed command, the data is migrated to the db. So if I feed the data as per custom attributes appearingin json file will theseed command feed the data to the db?

**Reply by AkshatKhamesra** [reply]
I wanted to ask the same @vholik , kindly help us on this :)

**Reply by AkshatKhamesra** [reply]
Hey @vholik , requesting your kind attention as we created a script to migrate data but the data won't show in the tables for the custom attributes and we would really appreciate if you have any tips for us on this. We are stuck on an important task that will be continued once this happens. Can you please guide us here?

Thanks in advance! :)

**Comment by modzincltd** [comment]
Thank you so much I will look at that right now

**Comment by bhavna82** [comment]
> I created the plugin that adds custom attributes to the products. If you're looking for a good solution definetely [check it out](https://www.npmjs.com/package/medusa-custom-attributes). Also there is a [blog post](https://rigby.pl/en/blog/custom-attributes-in-medusa.js) how to use that plugin. 

Can you please add support for latest Medusa version?
